### PR TITLE
Fix and refactor async code

### PR DESCRIPTION
**Pull request by RTLcoil**

### CODE REVIEW

File: Shared/Cloudinary.cs [hunk 1]
```diff
@@ -1304,6 +2759,29 @@ public RestoreResult Restore(RestoreParams parameters)
...
 
+        /// <summary>
+        /// Calls an upload mappings API.
+        /// </summary>
+        /// <param name="httpMethod">HTTP method.</param>
+        /// <param name="parameters">Parameters for Mapping of folders to URL prefixes for dynamic image fetching from
+        /// existing online locations.</param>
+        /// <param name="cancellationToken">(Optional) Cancellation token</param>
+        /// <returns>Parsed response after Upload mappings manipulation.</returns>
+        private Task<UploadMappingResults> CallUploadMappingsAPIAsync(HttpMethod httpMethod, UploadMappingParams parameters, CancellationToken? cancellationToken = null)
```
Comment (const-cloudinary) [inline Shared/Cloudinary.cs]: Rename to `CallUploadMappingsApiAsync`

File: Shared/Search.cs [hunk 2]
```diff
@@ -151,9 +153,35 @@ public Search SortBy(string field, string dir)
         /// <returns>Search response with information about the assets matching the search criteria.</returns>
         public SearchResult Execute()
         {
-            Url url = m_api.ApiUrlV.Add("resources").Add("search");
+            Url url = m_api.ApiUrlV
```
Comment (const-cloudinary) [inline Shared/Search.cs]: Extract to a property and reuse it

File: Shared/FileDescription.cs [hunk 3]
```diff
@@ -8,19 +8,37 @@ namespace CloudinaryDotNet
     /// </summary>
     public class FileDescription
     {
+        /// <summary>
+        /// Buffer length
```
[the remark this answers is not in the document]
Comment (RTLcoil) [reply Shared/FileDescription.cs]: In this case the name of this property is self describing, what comment would you suggest?

File: Shared/Cloudinary.cs [hunk 4]
```diff
@@ -5,14 +5,61 @@
...
+        private class UploadPresetApiParams
+        {
+            /// <summary>
+            /// Instantiates the <see cref="UploadPresetApiParams"/> object.
+            /// </summary>
+            /// <param name="httpMethod">Http request method.</param>
+            /// <param name="url">Url for api call.</param>
+            /// <param name="paramsCopy">Parameters of the upload preset.</param>
+            /// <param name="fileDescription">File description.</param>
+            public UploadPresetApiParams(HttpMethod httpMethod,
```
[the remark this answers is not in the document]
Comment (RTLcoil) [reply Shared/Cloudinary.cs]: This class is actually a tuple. It's purpose is to reduce the number of parameters that are the same in several method (https://refactoring.guru/introduce-parameter-object). It is a private one, we don't think it should be commented at all. 
We can go ahead and review the rest of API finding out other combination of parameters that are the same in different methods and propose how to restructure them in `XXXParams` objects.

File: Shared/Cloudinary.cs [hunk 5]
```diff
@@ -1320,6 +2868,23 @@ private UploadMappingResults CallUploadMappingsAPI(HttpMethod httpMethod, Upload
...
+        /// <param name="parameters">
+        /// Uses only <see cref="UploadMappingParams.MaxResults"/> and <see cref="UploadMappingParams.NextCursor"/>
+        /// properties. Can be null.
+        /// </param>
+        /// <param name="cancellationToken">(Optional) Cancellation token</param>
+        /// <returns>Parsed response after Upload mappings manipulation.</returns>
+        public Task<UploadMappingResults> UploadMappingsAsync(UploadMappingParams parameters, CancellationToken? cancellationToken = null)
+        {
+            if (parameters == null)
+                parameters = new UploadMappingParams();
```
[the remark this answers is not in the document]
Comment (RTLcoil) [reply Shared/Cloudinary.cs]: This code was copied from a sync version. There are `NullReferenceException` possible down the stack (`GetUploadMappingUrl(UploadMappingParams parameters)` for instance)

File: Shared/ApiShared.Internal.cs [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+
+        private StringContent CreateStringContent(SortedDictionary<string, object> parameters) =>
+            new StringContent(ParamsToJson(parameters), Encoding.UTF8, Constants.CONTENT_TYPE_APPLICATION_JSON);
+
+        private static bool IsStringContent(Dictionary<string, string> extraHeaders) =>
+            extraHeaders != null &&
+            extraHeaders.TryGetValue(Constants.HEADER_CONTENT_TYPE, out var value) &&
+            value == Constants.CONTENT_TYPE_APPLICATION_JSON;
+
+        private static void EnrichAndSetContent(HttpRequestMessage request, Dictionary<string, string> extraHeaders, HttpContent content)
```
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/ApiShared.Internal.cs]: How about: `SetHeadersAndContent`, if it's too specific, you can call it `FinalizeRequestContent`

File: Shared/FileDescription.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/FileDescription.cs]: @RTLcoil ,you check the docstring of the function that sets this value, this is actually a maximum size of a single chunk, used in uploadLarge

File: Shared/ApiShared.Internal.cs [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+            WriteLine(writer);
+            WriteLine(writer, value);
+        }
+
+        /// <summary>
+        /// Write cloudinary parameter to the request stream.
+        /// </summary>
+        /// <param name="writer">Stream writer.</param>
+        /// <param name="file">File to be written to the stream.</param>
+        protected void WriteFile(StreamWriter writer, FileDescription file)
```
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/ApiShared.Internal.cs]: @RTLcoil only public functions cannot be removed, no obligation to protected ones

File: Shared/Cloudinary.StreamingProfiles.cs [hunk 8]
```diff
@@ -66,5 +118,47 @@ public StreamingProfileListResult ListStreamingProfiles()
...
+            ValidateStreamingProfileUpdateParams(parameters);
+        }
+
+        private static void ValidateStreamingProfileUpdateParams(StreamingProfileUpdateParams parameters)
+        {
+            if (parameters == null)
+                throw new ArgumentNullException(nameof(parameters));
+        }
+
+        private static void ValidateNameForCallProfileApiParameters(string name)
```
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/Cloudinary.StreamingProfiles.cs]: yes

File: Shared/Cloudinary.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/Cloudinary.cs]: let's do it in a separate task

File: Shared/Cloudinary.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (const-cloudinary) [reply Shared/Cloudinary.cs]: It would be better to fix `NullReferenceException` there, or add FIXME here, to do it later

**Pull request merged**